src/libopenrave/kinbodygrab.cpp

```diff
@@ -118,17 +118,29 @@ Grabbed::Grabbed(KinBodyPtr pGrabbedBody, KinBody::LinkPtr pGrabbingLink)
     _pGrabbedBody = pGrabbedBody;
     _pGrabbingLink = pGrabbingLink;
     _pGrabbingLink->GetRigidlyAttachedLinks(_vAttachedToGrabbingLink);
+
+    // If the grabbing body has no joints, then it is effectively a rigid body.
+    // When calculating non-colliding links, we exclude all links rigidly attached to the grabbing link.
+    // Since for fully rigid bodies _all_ links are attached, we effectively exclude all links.
+    // We can therefore skip the non-colliding calculation, along with generating the associated state savers.
+    KinBodyPtr pGrabber(pGrabbingLink->GetParent());
+    if (pGrabber->GetJoints().size() + pGrabber->GetPassiveJoints().size() == 0) {
+        // This body is fully rigid: all links are excluded from the non colliding link set, so just flag that the list is valid (empty) and return.
+        _listNonCollidingIsValid = true;
+        return;
+    }
+
+    // if this body is not rigid, then we need to save the current state of the grabber/grabbee so that we can roll back to it when computing the non-colliding links
     _listNonCollidingIsValid = false;
     const bool bDisableRestoreOnDestructor = true; // This is very important! These saver are used only in ComputeListNonCollidingLinks and we don't want to restore on destructor.
+    static const int saverOptions = KinBody::Save_LinkTransformation | KinBody::Save_LinkEnable | KinBody::Save_JointLimits | KinBody::Save_ConnectedBodies;
     _CreateSaverForGrabbedAndGrabber(_pGrabbedSaver,
                                      pGrabbedBody,
-                                     KinBody::Save_LinkTransformation|KinBody::Save_LinkEnable|KinBody::Save_JointLimits,
+                                     saverOptions,
```

Comment:
If someone tries to set velocity of a currently grabbed kinbody, should we reject it because its velocity is determined by the grabbing link? If such guarantee exists that body's velocity doesn't get modified while being grabbed, it is fine  to not save at the time of grabbing and restore at the time of releasing. @rdiankov @rschlaikjer

Reply:
Should be okay if velocity is not saved..